### PR TITLE
[e2e] Resurrecting e2e tests on kind with envtest framework

### PR DESCRIPTION
**Pull request by xornivore**
### What does this PR do?

- Added initial tests for testing autoPause and autoFail of canary on pod restarts.

### Motivation

- While envtest control plane tests are great they don't allow us to exercise scenarios that assert on behaviors involving failure cases such as pod restarts.

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: controllers/suite_e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+func TestAPIs(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"Controller Suite",
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var _ = BeforeSuite(func(done Done) {
```
Comment (vboulineau) [inline controllers/suite_e2e_test.go:53]: Are we able to have that in common with the normal tests?

File: controllers/extendeddaemonset_e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+	const (
+		timeout     = 1 * time.Minute
+		longTimeout = 5 * time.Minute
+		interval    = 2 * time.Second
+	)
+
+	intString2 := intstr.FromInt(2)
+	intString10 := intstr.FromInt(10)
+
+	Context("Initial deployment", func() {
```
Comment (vboulineau) [inline controllers/extendeddaemonset_e2e_test.go:41]: I think we should have specific, dedicated tests if we are to introduce E2E with Cluster.

File: controllers/suite_e2e_test.go (hunk 1 above)
Comment (xornivore) [reply controllers/suite_e2e_test.go:53]: We can extract some parts here but there are differences:
- CRD versions (v1 does not install on k8s 1.19.1
- Use existing cluster boolean
- Node creation
I will try to deduplicate this a bit.

File: controllers/extendeddaemonset_e2e_test.go (hunk 2 above)
Comment (xornivore) [reply controllers/extendeddaemonset_e2e_test.go:41]: Do you mean splitting this monolithic one into separate tests? There is still a need for the initial deployment/canary phase before you can do anything else. Concern is duration of tests as well if we repeat that on every split test.

File: controllers/suite_e2e_test.go (hunk 1 above)
Comment (vboulineau) [reply controllers/suite_e2e_test.go:53]: V1 not installing is definitely an issue we should fix, it's supposed to work in 1.16+.
For the other two we should have some parameters or functors to customize setup

File: controllers/extendeddaemonset_e2e_test.go (hunk 2 above)
Comment (vboulineau) [reply controllers/extendeddaemonset_e2e_test.go:41]: I meant the duplication of tests, I believe these tests are duplicated from the original file:

```
It("Should handle EDS"
It("Should do canary deployment"
It("Should add canary labels"
It("Should remove canary labels"
```

Comment (xornivore) [reply controllers/extendeddaemonset_e2e_test.go:41]: Yes, so we can just wait for canary to be over. I will remove these.

**Review by clamoriniere — APPROVED** [review]
Awesome work @xornivore 🥇

File: controllers/extendeddaemonset_test.go [hunk 3]
```diff
@@ -57,145 +61,81 @@ var _ = Describe("ExtendedDaemonSet Controller", func() {
...
 				return len(canaryPods.Items) == 0
-			}, timeout, interval).Should(BeTrue())
+			}), timeout, interval).Should(BeTrue())
 		})
 	})
 
 	Context("Using ExtendedDaemonsetSetting", func() {
-		var err error
-		namespace := "default"
+		// var err error
```
Comment (clamoriniere) [inline controllers/extendeddaemonset_test.go:138]: cleanup 
```suggestion
```

**Pull request merged**